dataprofiler/tests/profilers/test_profile_builder.py

```diff
@@ -3790,6 +3801,279 @@ def test_null_row_stats_correct_after_updates(self, *mocks):
+
+    def test_unique_row_ratio_empty_profiler(self):
+        profiler_options = ProfilerOptions()
+        profiler_options.set(
+            {
+                "*.is_enabled": False,
+                "row_statistics.is_enabled": True,
+            }
+        )
+        profiler = StructuredProfiler(pd.DataFrame([]), options=profiler_options)
+        self.assertEqual(0, profiler._get_unique_row_ratio())
+
+    def test_correct_duplicate_row_count_full_row_hashing(self):
+        self.assertEqual(110, len(self.trained_schema_full.hashed_row_object))
+        self.assertEqual(200, self.trained_schema_full.total_samples)
+        self.assertEqual(90, self.trained_schema_full._get_duplicate_row_count())
+
+    def test_correct_duplicate_row_count_hll_row_hashing(self):
+        # if data changes, hll might not be completely accurate since it is estimation
+        self.assertEqual(110, self.trained_schema_hll.hashed_row_object.cardinality())
+        self.assertEqual(200, self.trained_schema_hll.total_samples)
+        self.assertEqual(90, self.trained_schema_hll._get_duplicate_row_count())
+
+    def test_correct_duplicate_row_count_unique_count_disabled(self):
+        profiler_options_1 = ProfilerOptions()
+        profiler_options_1.set(
+            {
+                "*.is_enabled": False,
+            }
+        )
+        profiler = StructuredProfiler(pd.DataFrame([]), options=profiler_options_1)
+        self.assertIsNone(profiler._get_duplicate_row_count())
+
+    def test_correct_duplicate_row_count_empty_profiler(self):
+        profiler_options = ProfilerOptions()
+        profiler_options.set(
+            {
+                "*.is_enabled": False,
+                "row_statistics.is_enabled": True,
+            }
+        )
+        profiler = StructuredProfiler(pd.DataFrame([]), options=profiler_options)
+        self.assertEqual(0, profiler._get_duplicate_row_count())
+
+    def test_duplicate_row_count_hll_cardinality_greater_than_total_samples(self):
+        profiler_options = ProfilerOptions()
+        profiler_options.set(
+            {
+                "*.is_enabled": False,
+                "row_statistics.is_enabled": True,
+                "row_statistics.unique_count.is_enabled": True,
+                "row_statistics.unique_count.hashing_method": "hll",
+            }
+        )
+        with mock.patch(
+            "dataprofiler.profilers.profile_builder.HyperLogLog",
+            spec=dataprofiler.profilers.profile_builder.HyperLogLog,
+        ) as hll_mock:
+            hll_mock.return_value.cardinality.return_value = 1000
+            profiler = StructuredProfiler(pd.DataFrame([]), options=profiler_options)
+
+        self.assertEqual(1000, profiler.hashed_row_object.cardinality())
+        self.assertEqual(0, profiler._get_duplicate_row_count())
+
+    def test_save_and_load_hll(self):
+        self.assertEqual(110, self.trained_schema_hll.hashed_row_object.cardinality())
```

Comment:
This feels a little strange to test as opposed to just comparing before and after to each other.